### PR TITLE
Bind mounts should be (re)created when request uses container name

### PR DESCRIPTION
**Pull request by wolfymaster**
This change adds support for re-binding mounts when starting/stopping/restarting containers using the container's name.

One caveat is that the container must have been explicitly assigned a name at creation and not randomly assigned a name by docker. 

### UPDATE 2022-12-31:

Based on the feedback provided, I have updated this PR with a different approach that satisfies all conditions. For any proxied request, we use the provided `id` path template variable to make an upstream request to the docker engine api to inspect the container. Fortunately it supports *both* id or name as the container identifier. The `Id` returned will be the full long container id that is used to lookup in docker-binds.json. We replace the `template["id"]` value with the returned `Id` so that any mungers always receive the full container id. 

Fixes: https://github.com/rancher-sandbox/rancher-desktop/issues/2231 

Signed-off-by: Paul Sherer <wolfy@wolfymaster.com>

### CODE REVIEW

**Review by mook-as — CHANGES_REQUESTED** [review]
Thank you for digging into the issue and submitting a PR!

Unfortunately, I don't think this quite fixes the issue: it's possible to refer to the container by using a unique prefix (IIRC) of the container id:

```
$ mkdir vol
$ date > vol/hello.txt
$ docker run -itd --restart=always -v $PWD/vol:/vol --name cc alpine
ccb7ffe80e2b85244ef2e3b85b24ec85333f73de26bee9ffd6fc93ab8b58c25d
$ docker exec -ti cc cat /vol/hello.txt
Wed 30 Nov 2022 01:55:22 PM PST
$ docker stop cc
cc
$ docker start cc
cc
$ docker exec -i cc cat /vol/hello.txt
Wed 30 Nov 2022 01:55:22 PM PST
$ docker stop cc
cc
$ docker ps -a
CONTAINER ID   IMAGE     COMMAND     CREATED         STATUS                       PORTS     NAMES
ccb7ffe80e2b   alpine    "/bin/sh"   4 minutes ago   Exited (137) 4 seconds ago             cc
$ docker start ccb7ffe80e2b
ccb7ffe80e2b
$ docker exec -i cc cat /vol/hello.txt
cat: can't open '/vol/hello.txt': No such file or directory
```

I think we'll have to [inspect the container](https://docs.docker.com/engine/api/v1.41/#tag/Container/operation/ContainerInspect) to convert the input string to the full container ID, and then look up the mount information via that full ID. Does that sound reasonable to you?

**Comment by wolfymaster** [comment]
@mook-as Thanks for the feedback! Yes, I must admit that I was also aware of this and that the PR does not handle that case. A more complete solution is to do exactly what you stated and in a way, "resolve" the identifier provided to the full container id. My hesitancy toward that approach was the size of the change required to fix the original specific issue. 

I'm happy to revise this PR to take a stab at implementing that complete solution.

**Comment by mook-as** [comment]
> I'm happy to revise this PR to take a stab at implementing that complete solution.

That sounds great — I'll monitor the PR† and re-review when I see the changes!

†Actually I just watch the whole repo so I get every message, but you get the idea :)

**Comment by ryancurrah** [comment]
We just adopted Rancher Desktop at our organization and some of our Windows developers are running into this issue. Look forward to seeing this in the next release! Thank you.

**Review by mook-as — CHANGES_REQUESTED** [review]
Looks good, just some minor comments.

File: src/go/wsl-helper/pkg/dockerproxy/serve.go [hunk 1]
```diff
@@ -227,6 +255,46 @@ func (m *requestMunger) MungeResponse(resp *http.Response) error {
...
+func (m *requestMunger) InspectContainerRequest(req *http.Request, id string, dialer func() (net.Conn, error)) (*containerInspectResponseBody, error) {
+	// extract version from request
+	match := m.apiDetectPattern.FindStringIndex(req.URL.Path)
+	version := "/"
+	if match != nil {
+		version = req.URL.Path[match[0]:match[1]]
+	}
+
+	// url for inspecting container
+	inspectURL := fmt.Sprintf("%s://%s%scontainers/%s/json", req.URL.Scheme, req.URL.Host, version, id)
```
[the remark this answers is not in the document]
Comment (wolfymaster) [reply src/go/wsl-helper/pkg/dockerproxy/serve.go:267]: I'd be fine with hardcoding a value. Do you have an opinion on where that should be declared? I'm not a fan of hardcoding within the string.

File: src/go/wsl-helper/pkg/dockerproxy/serve.go [hunk 2]
```diff
@@ -227,6 +255,46 @@ func (m *requestMunger) MungeResponse(resp *http.Response) error {
 	return nil
 }
 
+func (m *requestMunger) InspectContainerRequest(req *http.Request, id string, dialer func() (net.Conn, error)) (*containerInspectResponseBody, error) {
```
[the remark this answers is not in the document]
Comment (wolfymaster) [reply src/go/wsl-helper/pkg/dockerproxy/serve.go:258]: ah.. yes. I understand the confusion. I like `CanonicalizeContainerID`. Thanks for that feedback! I appreciate it!

File: src/go/wsl-helper/pkg/dockerproxy/serve.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mook-as) [reply src/go/wsl-helper/pkg/dockerproxy/serve.go:267]: Hmm, just as a `const` within the file? I don't expect us to want to change it very often.

**Review by mook-as — COMMENTED** [review]
Forgot to mention it before — thanks a lot for working on this!

**Comment by mook-as** [comment]
Hi @wolfymaster! Just wanted to check how things are going. This PR is pretty close (just down to trivial changes), and we didn't want to miss the momentum :)

**Comment by wolfymaster** [comment]
@mook-as yeah, I almost have everything buttoned up.

**Comment by mook-as** [comment]
The E2E tests are failing because of:
```
2023-01-20T19:08:24.379Z: Kubernetes was unable to start: Error: /home/ranchertest/src/resources/linux/lima/bin/limactl exited with code 1
    at ChildProcess.<anonymous> (/home/ranchertest/src/dist/app/background.js:56096:27)
    at ChildProcess.emit (node:events:390:28)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12) {
  stdout: '',
  stderr: 'time="2023-01-20T19:08:24Z" level=debug msg="interpreting argument \\"0\\" as an instance name"\n' +
    'time="2023-01-20T19:08:24Z" level=info msg="Using the existing instance \\"0\\""\n' +
    'time="2023-01-20T19:08:24Z" level=info msg="Attempting to download the image from \\"/home/ranchertest/src/resources/linux/alpine-lima-v0.2.25.rd2-rd-3.16.0.iso\\"" digest=\n' +
    'time="2023-01-20T19:08:24Z" level=fatal msg="failed to download the image, attempted 1 candidates, errors=[failed to download \\"/home/ranchertest/src/resources/linux/alpine-lima-v0.2.25.rd2-rd-3.16.0.iso\\": failed to open source /home/ranchertest/src/resources/linux/alpine-lima-v0.2.25.rd2-rd-3.16.0.iso: open /home/ranchertest/src/resources/linux/alpine-lima-v0.2.25.
```
Looking at the playwright trace logs, there _is_ one of our error dialogs.

But the build log contains:
```
Downloading https://github.com/rancher-sandbox/alpine-lima/releases/download/v0.2.25.rd2/alpine-lima-rd-3.16.0-x86_64.iso to /home/ranchertest/src/resources/linux/alpine-lima-v0.2.25.rd2-rd-3.16.0.iso...
```
So the file should have been downloaded. But I also notice in the diagnostics logs:
```
2023-01-20T19:08:24.190Z: Check RD_BIN_DOCKER_CLI_SYMLINK_DOCKER-COMPOSE.DOWNLOAD result: {"description":"The file `~/.rd/bin/docker-compose.download` is a symlink to `~/src/resources/linux/bin/docker-compose.download`, which is not executable.","passed":false,"fixes":[]}
```
So that sounds like we started downloading the file, and was somehow aborted without even doing the `finally` bit to to clean up, somewhere around https://github.com/rancher-sandbox/rancher-desktop/blob/8352cbf3877986e0ba44fdf5dd132e08e254b248/scripts/lib/download.ts#L75-L99

I _think_ we're good to override the CI here, but let me discuss with some of the other folks.

**Comment by mook-as** [comment]
Discussed offline, we're going to merge this as-is, since the affected files shouldn't even be relevant for the tests that failed. (The changed code is Windows-only, and the failing test runs are only on Linux.)

**Comment by mook-as** [comment]
Merged, thanks!

I've slapped the `release-note` label on in case we want to highlight external contribution in the release notes (I'm not actually what the normal thing there is, so that's not a guarantee that it will actually land in the release note).

**Pull request merged**